### PR TITLE
Use DAA score where needed

### LINKED ISSUE #1259: DAA score and minting kaspa
The DAA gives is 1b/s only locally, meaning that if we want to measure time in the system we have to track blocks that were in the DAA window, because only they have affected the difficulty.

For that we can inherit the "DAA Score" from the SP and add blocks any block from the mergeSet that is in the blueWindow of the block.
We also need to mark those block for kaspa rewards, that way we can control the inflation rate.

This is the same as block height under k=0 and rejecting red blocks (bitcoin).

 Note that according to #1029 we will only give reward to blocks that have affected the DAA, and obviously the reward of red blocks goes to the merging block according to #1029

### LINKED ISSUE #1276: Bestow the rewards and fees of red blocks to its merging block
The rewards of red blocks (fees + CB txn) should go to the merging block, rather than the actual block, more on that should be detailed in the monetary policy issue.

For more info: #1029

### PR DESCRIPTION
**Pull request by someone235**
* Usee DAA score for UTXO entries
* Use DAA score for coinbase maturity.
* Use DAA score for sequence lock
* Calculate halvings by DAA score
* Don't pay to blocks that are not included in the DAA added blocks, and bestow red blocks reward to the merging block.

This closes #1259 and closes #1276

### CODE REVIEW

File: cmd/wallet/common.go [hunk 1]
```diff
@@ -10,7 +10,8 @@ func isUTXOSpendable(entry *appmessage.UTXOsByAddressesEntry, virtualSelectedPar
 	if !entry.UTXOEntry.IsCoinbase {
 		return true
 	}
-	blockBlueScore := entry.UTXOEntry.BlockBlueScore
+	blockBlueScore := entry.UTXOEntry.BlockDAAScore
+	// TODO: Check for a better alternative than virtualSelectedParentBlueScore
```
Comment (svarogg) [inline cmd/wallet/common.go:14]: Do we have an open issue for this?

File: domain/consensus/processes/blockbuilder/test_block_builder.go [hunk 2]
```diff
@@ -136,6 +136,11 @@ func (bb *testBlockBuilder) buildBlockWithParents(parentHashes []*externalapi.Do
 		return nil, nil, err
 	}
 
+	_, err = bb.difficultyManager.UpdateDAADataAndReturnDifficultyBits(tempBlockHash)
+	if err != nil {
+		return nil, nil, err
+	}
+
```
Comment (elichai) [inline domain/consensus/processes/blockbuilder/test_block_builder.go:143]: Why is this needed?

File: domain/consensus/processes/blockbuilder/test_block_builder.go [hunk 3]
```diff
@@ -197,6 +202,11 @@ func (bb *testBlockBuilder) BuildUTXOInvalidBlock(parentHashes []*externalapi.Do
 		return nil, err
 	}
 
+	_, err = bb.difficultyManager.UpdateDAADataAndReturnDifficultyBits(tempBlockHash)
+	if err != nil {
+		return nil, err
+	}
+
```
Comment (elichai) [inline domain/consensus/processes/blockbuilder/test_block_builder.go:209]: Same

File: domain/consensus/consensus.go [hunk 4]
```diff
@@ -313,7 +313,7 @@ func (s *consensus) GetVirtualInfo() (*externalapi.VirtualInfo, error) {
 	if err != nil {
 		return nil, err
 	}
-	bits, err := s.difficultyManager.RequiredDifficulty(model.VirtualBlockHash)
+	bits, err := s.difficultyManager.UpdateDAADataAndReturnDifficultyBits(model.VirtualBlockHash)
```
Comment (svarogg) [inline domain/consensus/consensus.go:316]: Why do we update anything in a GetXXX function?

File: domain/consensus/processes/coinbasemanager/coinbasemanager.go [hunk 5]
```diff
@@ -58,25 +74,30 @@ func (c *coinbaseManager) ExpectedCoinbaseTransaction(blockHash *externalapi.Dom
 	}, nil
 }
 
+func (c *coinbaseManager) daaAddedBlocksSet(blockHash *externalapi.DomainHash) (hashset.HashSet, error) {
+	daaAddedBlocks, err := c.daaBlocksStore.DAAAddedBlocks(c.databaseContext, blockHash)
+	if err != nil {
+		return nil, err
+	}
+
+	daaAddedBlocksSet := hashset.New()
```
Comment (elichai) [inline domain/consensus/processes/coinbasemanager/coinbasemanager.go:83]: You can just call `hashset.NewFromSlice(daaAddedBlocks)` (which should probably be changed to preallocate)

File: domain/consensus/processes/blockbuilder/test_block_builder.go [hunk 6]
```diff
@@ -136,6 +136,11 @@ func (bb *testBlockBuilder) buildBlockWithParents(parentHashes []*externalapi.Do
 		return nil, nil, err
 	}
 
+	_, err = bb.difficultyManager.UpdateDAADataAndReturnDifficultyBits(tempBlockHash)
```
Comment (svarogg) [inline domain/consensus/processes/blockbuilder/test_block_builder.go:139]: Here too, I'd expect this to get updated every time we add a block, then retrieved by buildBlockWithParents, not calculated every time again and again.

Comment (svarogg) [reply domain/consensus/processes/blockbuilder/test_block_builder.go:139]: Alternatively, since this is calculated each and every time it's accessed, maybe we just shouldn't store it?

File: infrastructure/network/netadapter/server/grpcserver/protowire/p2p.proto [hunk 7]
```diff
@@ -161,7 +161,7 @@ message OutpointAndUtxoEntryPair{
 message UtxoEntry {
   uint64 amount = 1;
   ScriptPublicKey scriptPublicKey = 2;
-  uint64 blockBlueScore = 3;
+  uint64 blockDaaScore = 3;
```
Comment (svarogg) [inline infrastructure/network/netadapter/server/grpcserver/protowire/p2p.proto:164]: You have places where you spell Daa and places where you spell DAA, I think all should be DAA

File: domain/consensus/processes/coinbasemanager/coinbasemanager.go [hunk 8]
```diff
@@ -95,6 +116,31 @@ func (c *coinbaseManager) coinbaseOutputForBlueBlock(blueBlock *externalapi.Doma
...
 }
 
+func (c *coinbaseManager) coinbaseOutputForRewardFromRedBlocks(ghostdagData *model.BlockGHOSTDAGData,
+	acceptanceData externalapi.AcceptanceData, daaAddedBlocksSet hashset.HashSet,
+	coinbaseData *externalapi.DomainCoinbaseData) (*externalapi.DomainTransactionOutput, bool, error) {
+
+	totalReward := uint64(0)
+	mergeSetBluesCount := len(ghostdagData.MergeSetBlues())
+	for i, red := range ghostdagData.MergeSetReds() {
+		reward, err := c.calcMergedBlockReward(red, acceptanceData[mergeSetBluesCount+i], daaAddedBlocksSet)
```
Comment (elichai) [inline domain/consensus/processes/coinbasemanager/coinbasemanager.go:126]: We need to be careful with this, if we ever change the ordering of the acceptance data(ie https://github.com/kaspanet/kaspad/issues/1558)

**Review by svarogg — CHANGES_REQUESTED** [review]

File: cmd/wallet/common.go (hunk 1 above)
Comment (someone235) [reply cmd/wallet/common.go:14]: Yes https://github.com/kaspanet/kaspad/issues/1599

File: domain/consensus/consensus.go (hunk 4 above)
Comment (someone235) [reply domain/consensus/consensus.go:316]: It's the same GHOSTDAG pattern

File: domain/consensus/processes/blockbuilder/test_block_builder.go (hunk 6 above)
Comment (someone235) [reply domain/consensus/processes/blockbuilder/test_block_builder.go:139]: The DAA score is calculated incrementally, so we must store it

File: infrastructure/network/netadapter/server/grpcserver/protowire/p2p.proto (hunk 7 above)
Comment (someone235) [reply infrastructure/network/netadapter/server/grpcserver/protowire/p2p.proto:164]: AFAIK the convention in proto files is to use js-like camel case

File: domain/consensus/processes/consensusstatemanager/update_virtual.go [hunk 9]
```diff
@@ -42,6 +42,11 @@ func (csm *consensusStateManager) updateVirtual(newBlockHash *externalapi.Domain
 		return nil, nil, err
 	}
 
+	_, err = csm.difficultyManager.UpdateDAADataAndReturnDifficultyBits(model.VirtualBlockHash)
+	if err != nil {
+		return nil, nil, err
+	}
+
```
Comment (elichai) [inline domain/consensus/processes/consensusstatemanager/update_virtual.go:49]: Is this needed for the `CalculatePastUTXOAndAcceptanceData`?

File: domain/consensus/processes/consensusstatemanager/resolve_block_status_test.go [hunk 10]
```diff
@@ -261,7 +261,12 @@ func TestTransactionAcceptance(t *testing.T) {
 			t.Fatalf("Error creating tip: %+v", err)
 		}
 
-		finalTipHash, _, err := testConsensus.AddBlock([]*externalapi.DomainHash{finalTipSelectedParentHash, redHash, blueHash}, nil,
+		finalTipScriptPublicKey := &externalapi.ScriptPublicKey{Script: []byte{4}, Version: 0}
```
Comment (elichai) [inline domain/consensus/processes/consensusstatemanager/resolve_block_status_test.go:264]: What changed?

File: domain/consensus/processes/consensusstatemanager/update_virtual.go (hunk 9 above)
Comment (someone235) [reply domain/consensus/processes/consensusstatemanager/update_virtual.go:49]: Yes

File: domain/consensus/processes/consensusstatemanager/resolve_block_status_test.go (hunk 10 above)
Comment (someone235) [reply domain/consensus/processes/consensusstatemanager/resolve_block_status_test.go:264]: The red block reward is bestowed to the merging block instead of disappearing

File: domain/consensus/processes/blockbuilder/test_block_builder.go (hunk 2 above)
Comment (someone235) [reply domain/consensus/processes/blockbuilder/test_block_builder.go:143]: This is a prerequisite for ResolveBlockStatus

File: domain/consensus/consensus.go (hunk 4 above)
Comment (svarogg) [reply domain/consensus/consensus.go:316]: I think I understand what you mean, but not 100%, but forget about explaining me, this needs a lengthy explanation in the comment for UpdateDAADataAndReturnDifficultyBits, explaining why do we need to update it every time we return it? Why is it done incrementally? And what "being done incrementally" exactly means in this context.

File: domain/consensus/processes/difficultymanager/difficultymanager.go [hunk 11]
```diff
@@ -67,19 +67,40 @@ func (dm *difficultyManager) genesisBits() (uint32, error) {
 	return header.Bits(), nil
 }
 
-// RequiredDifficulty returns the difficulty required for some block
-func (dm *difficultyManager) UpdateDAADataAndReturnDifficultyBits(blockHash *externalapi.DomainHash) (uint32, error) {
+// StageDAADataAndReturnRequiredDifficulty calculates the DAA window, stages the DAA score and DAA added
+// blocks, and returns the required difficulty for the given block.
+// The reason this function both stages DAA data and returns the difficulty is because in order to calculate
+// both of them we need to calculate the DAA window, which is a relatively heavy operation, so reuse the block
```
Comment (svarogg) [inline domain/consensus/processes/difficultymanager/difficultymanager.go:73]: Grammar: so **we** reuse

File: domain/consensus/processes/difficultymanager/difficultymanager.go [hunk 12]
```diff
@@ -67,19 +67,40 @@ func (dm *difficultyManager) genesisBits() (uint32, error) {
...
+// window instead of recalculating it for the two purposes.
+// For cases where no staging should be happen and the caller only needs to know the difficulty RequiredDifficulty
```
Comment (svarogg) [inline domain/consensus/processes/difficultymanager/difficultymanager.go:75]: Grammar: no staging should **be** happen

I'd also rephrase "... the caller only needs to know the difficulty RequiredDifficulty should be used" to "... the caller only needs to know the difficulty - use RequiredDifficulty instead". (If you disagree at least add some separator before RequiredDifficulty)

**Review by svarogg — CHANGES_REQUESTED** [review]

**Review by svarogg — APPROVED** [review]

File: domain/consensus/processes/blockbuilder/test_block_builder.go [hunk 13]
```diff
@@ -199,7 +199,13 @@ func (bb *testBlockBuilder) BuildUTXOInvalidBlock(parentHashes []*externalapi.Do
 
 	// We use genesis transactions so we'll have something to build merkle root and coinbase with
 	genesisTransactions := bb.testConsensus.DAGParams().GenesisBlock.Transactions
-	header, err := bb.buildUTXOInvalidHeader(parentHashes, genesisTransactions)
+
+	bits, err := bb.difficultyManager.RequiredDifficulty(tempBlockHash)
```
Comment (elichai) [inline domain/consensus/processes/blockbuilder/test_block_builder.go:203]: I think you meant `StageDAADataAndReturnRequiredDifficulty`

Comment (someone235) [reply domain/consensus/processes/blockbuilder/test_block_builder.go:203]: No, there's no need to know the DAA score of a block in order to build a UTXO invalid block

**Review by elichai — APPROVED** [review]
